STN-675: WEST Gradient Hero --> configuration

heatherdesigns opened the pull request:
# [STN-675](https://sparkbox.atlassian.net/browse/STN-675) READY FOR REVIEW

## Summary
The work in this PR creates a Gradient Hero component that allows to highlight imagery in the Hero Region and Components region on a page. This PR focuses on creating the new gradient hero paragraph type, styles are not included.

_Note:_ Timeline and timeline item are referenced in the configuration files. I did not remove them because I assumed that the work on the timeline will probably be merged before this PR is. If that is not the case, I can remove any references to timeline.

## Need Review By (Date)
asap

## Urgency
medium

## Steps to Test
1. In your CLI, run npm run test and confirm all tests pass.
2. In order for the gradient hero paragraph to be recognized, you will need to clear your cache by running `lando drush @sparkbox_sandbox.local cr`. Run `lando drush @sparkbox_sandbox.local config-import --partial` to import the config files.
3. To test a gradient, create a new flexible page or open an existing one. Under Components add a Gradient Hero. Fill in all the fields.
4. By default, the gradient hero should not be available via the components dropdown when creating a flexible page. In order to test the work in this PR you will need to go to the Component Settings for a Flexible Page and `uncheck` the `Gradient Hero` under `Reference Type`. Be sure to revert this change when you are done testing the scenarios listed below. Quick link:  http://sparkbox-sandbox.suhumsci.loc/admin/structure/types/manage/hs_basic_page/fields/node.hs_basic_page.field_hs_page_components
5. By default, the gradient hero should also not be available via the row component dropdown. In order to test the gradient hero within a row, you will need to go to the Component Settings for a Flexible Page for Rows and `uncheck` the `Gradient Hero` under `Reference Type`. Again be sure to revert this change when testing is complete. Quick link: http://sparkbox-sandbox.suhumsci.loc/admin/structure/paragraphs_type/hs_row/fields
6. To test the gradient hero in the hero section of a flexible page, you must add it to the hero image settings configuration. Under this quick link: http://sparkbox-sandbox.suhumsci.loc/admin/structure/types/manage/hs_basic_page/fields/node.hs_basic_page.field_hs_page_hero `check` the `Gradient Hero` under `Reference Type`. Remember to revert this change when testing is complete.

Test the following scenarios:
- [x] A gradient hero within the hero component section at the top of a page.
- [x] A gradient hero within the main content section of a page.
- [x] The gradient hero contains the following fields: image, title, body, link, and align
- [x] A gradient hero with default alignment to the left. Open your browser's dev tools and confirm that the class of `hb-hero-gradient--content-default` co-exists beside the pattern's class of `hb-gradient-hero`.
- [x] A gradient hero where the alignment has been changed to `-None-` also has the `hb-hero-gradient--content-default` class.
- [x] A gradient hero where the alignment has been set to `Right Aligned` should have the class of `hb-hero-gradient--content-right`.
- [x] Add a caption / credit to a gradient hero image. It should display the correct caption / credit styles (white text on a black background that fades to transparent.)
- [x] The gradient hero is keyboard accessible.
- [x] The addition of the gradient hero paragraph type does not affect any of the existing paragraph types.
- [x] The gradient hero can be added to the Colorful and Traditional themes.

## Expected results of adding a gradient hero to a flexible page. Note there are no styles.
<img width="1354" alt="gradient hero" src="https://user-images.githubusercontent.com/12678977/113316692-a09c7280-92dc-11eb-9c06-35827b02164d.png">

Commit:
feat (STN-675): West Gradient Hero configuration

Commit:
feat: add configuration files

Commit:
feat: update configuration so that captions/credits display on gradient hero images

Commit:
feat: update paragraph and row component config files so that paragraph types are listed alphabetically

Commit:
feat: update codeception test for the gradient hero paragraph type

Commit:
feat: update configuration files to hide the gradient hero from components by default

Commit:
feat: update gradient hero paragraph entity view display configuration so that the hero title displays as an h2 tag

heatherdesigns commented on docroot/themes/humsci/humsci_basic/patterns/gradient-hero/gradient-hero.html.twig:
```diff
@@ -0,0 +1,23 @@
+
+{% spaceless %}
+  <div{{ attributes }}>
+    <div class="hb-gradient-hero__wrapper">
```
As I was doing the styles, I found that this `div` with the `hb-gradient-hero__wrapper` class unnecessary. So it is removed in `STN-675--gradient-hero-styles` branch.

heatherdesigns commented:
Note to self: Body text needs to be limited. [FIXED!!]

I noticed it is not limited when working on the styles:
<img width="648" alt="Screen Shot 2021-04-07 at 3 17 15 PM" src="https://user-images.githubusercontent.com/12678977/113921890-984ca780-97b4-11eb-8786-23c2899a9262.png">

heatherdesigns commented:
[FIXED!!]
Another note to self: Images appear to be scaling differently between the hero image overly paragraph type and the gradient hero paragraph type. Compare config files and see if something is missing from gradient hero configs.

![image scaling](https://user-images.githubusercontent.com/12678977/113922392-3d678000-97b5-11eb-87fb-9addb7047a25.png)

heatherdesigns commented on config/default/field.field.node.hs_basic_page.field_hs_page_hero.yml:
```diff
@@ -60,13 +63,25 @@ settings:
+        enabled: false
+      hs_timeline_item:
+        weight: 31
+        enabled: false
```
This file probably wasn't included in the timeline work and probably should have been.

heatherdesigns commented on config/default/field.field.paragraph_row.hs_flexible_page.field_hs_flexible_components.yml:
```diff
@@ -58,13 +61,25 @@ settings:
+        enabled: false
+      hs_timeline_item:
+        weight: 31
+        enabled: false
```
Note to self. Check to see if this file was included in the timeline work. Because this makes it look like it wasn't.

heatherdesigns replied:
Remove everything that isn't hs_gradient_hero. See latest config on the release branch: https://github.com/SU-HSDO/suhumsci/blob/926300c95dbf7ac1dc5ac13c9efb34f8ff9001dd/config/default/field.field.paragraph_row.hs_flexible_page.field_hs_flexible_components.yml

heatherdesigns commented on config/default/field.field.paragraph_row.hs_flexible_page.field_hs_flexible_components.yml:
```diff
@@ -58,13 +61,25 @@ settings:
       hs_webform:
         weight: 24
         enabled: false
+      stanford_gallery:
```
Note to self: Is this the new photo grid????

Commit:
feat: update components configs to match the release and sprint branches before the addition of the gradient hero

Commit:
feat: Gradient hero configuration updates

- Add smart trim to body
- Update image styles to the correct setting which matches the hero image overlay component

kaseybon commented:
I don't think the default variant class is going to be necessary. We can style the left variant as the default for the class `hb-gradient-hero` and then right align the contents for the right variant.

That being said, I like how Ryan handled the left/right options on the callout component. Can we remove the none option and just have the default alignment set to left?

Commit:
feat: address review feedback

- require variant conditional field to match right variant to prevernt adding a variant class to the default hero
- Update the hero alignment configuration to make the alignment a required field. This elminates -None- as an option for users in the alignment dropdown.

rkelbel commented on tests/codeception/acceptance/Install/Content/FlexiblePageCest.php:
```diff
@@ -14,7 +14,7 @@ public function testPostCard(AcceptanceTester $I) {
     $I->amOnPage('/node/add/hs_basic_page');
     $I->canSeeNumberOfElements('#edit-field-hs-page-hero-wrapper', 1);
-    $I->canSeeNumberOfElements('#edit-field-hs-page-components-add-more input', 11);
+    $I->canSeeNumberOfElements('#edit-field-hs-page-components-add-more input', 12;
```
I believe this needs to go up to 13 now.

rkelbel replied:
Resolved! 🙂

Commit:
feat: update number of components to be tested

Commit:
feat: fix syntax in failing test

Commit:
feat: update number of elements tested

Commit:
feat: update gradient hero form display to show text area character limit feedback

rkelbel reviewed (approved):
Awesome, config looks great!
🙂
